dnf-testing.py

```diff
@@ -0,0 +1,351 @@
+    parser.add_argument(
+        "command", metavar="<command>",
+        choices=['list', 'build', 'run', 'shell'],
+        help="list: List of available functional tests\n"
+             "build: Build container with functional tests\n"
+             "run: Run all tests. The set of tests can be optionally specified by [TEST...]\n"
+             "shell: Run a bash shell session within the container"
+    )
+    parser.add_argument("args", metavar="<arg>", nargs="*")
+    return parser
+
+
+def error(msg):
+    print('Error: ' + msg, file=sys.stderr)
+    sys.exit(1)
+
+
+def execute(command):
+    '''
+    Execute the `command` in a new process, wait for process to finish and
+    return (exit code, [captured sdout lines], [captured stderr lines])
+    '''
+    #print('Running: ', ' '.join(command), file=sys.stderr)
+    process = subprocess.Popen(
+        command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    out_encoding = sys.stdout.encoding or 'utf-8'
+    stdout, stderr = process.communicate()
+    return process.returncode,\
+           stdout.decode(out_encoding).split('\n'),\
+           stderr.decode(out_encoding).split('\n')
+
+
+class BehaveRunner(object):
+    BUILD_TYPES = ['jjb', 'local', 'side-tag', 'distro']
+
+    def __init__(self):
+        self.command_line_parser = command_line_parser()
+        self.command_line_args = self.command_line_parser.parse_args()
+        self._docker_bin = None
+        self._build_type = None
+        self._volumes = None
+
+    def __call__(self):
+        command = getattr(self, 'command_%s' % self.command_line_args.command, None)
+        if not command:
+            error("Command '{}' is not implemented.".format(
+                self.command_line_args.command))
+        #print(self.command_line_args)
+        command()
+
+    @property
+    def param_reserve(self):
+        if self.command_line_args.reserve:
+            return ['-r']
+        elif self.command_line_args.reserveonfail:
+            return ['-R']
+        else:
+            return []
+
+    @property
+    def param_tty(self):
+        retval = []
+        if self.param_reserve:
+            retval.append('-it')
+        return retval
+
+    @property
+    def param_cache(self):
+        retval = []
+        if not self.command_line_args.usecache:
+            retval.append('--no-cache')
+        return retval
+
+    @property
+    def param_dnfcommand(self):
+        retval = []
+        if self.command_line_args.dnf_command:
+            retval.extend(['--command', self.command_line_args.dnf_command])
+        return retval
+
+    @property
+    def docker_bin(self):
+        '''
+        Autodetection whether podman or docker is gonna be used
+        '''
+        if self._docker_bin is None:
+            if self.command_line_args.podman:
+                self._docker_bin = 'podman'
+            else:
+                if execute(['command', '-v', 'docker'])[0] \
```

Comment:
Does this work for you? `command` is a bash built-in and you need to call `subprocess.Popen(..., shell=True)` for this to work. If I run this script with `-p` (telling it to use podman), it works, without `-p` I get a traceback on this. After fixing this one I'm getting another traceback when trying to run with docker, please test it and if it works for you, we can discuss privately so that we don't spam here...

(my podman is currently acting up, I can only test with docker for now)

Reply:
Yes, it does. But after some digging on internet I found this: https://unix.stackexchange.com/questions/429400/command-command-not-found-in-python-subprocess
Is this the traceback you are getting? I tried to fix.

Reply:
I don't remember any more, but it's working for me now.

Reply:
Actually, with `--docker`, it doesn't:
```
Traceback (most recent call last):
  File "./dnf-testing.py", line 369, in <module>
    BehaveRunner()()
  File "./dnf-testing.py", line 134, in __call__
    command()
  File "./dnf-testing.py", line 363, in command_build
    returncode = subprocess.call(command)
  File "/usr/lib/python3.7/subprocess.py", line 339, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/usr/lib/python3.7/subprocess.py", line 800, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.7/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'sudo docker': 'sudo docker'
```

The issue is with using `"sudo docker"` as the binary for `subprocess.call()`, if I replace it with two items in the list: `["sudo", "docker", ...]`, it works.